File: MtspSolver.cs
using System;
using System.Collections.Generic;

namespace mtsp
{
    public class MtspSolver
    {
        // ========================
        public static int MaximumSameResults = 1000;
        public static int InitialNumber = 16;
        public static int CrossoverNumber = InitialNumber;
        public static int MutationNumber = InitialNumber;
        public static int SelectionGroupSize =
            (InitialNumber + CrossoverNumber + MutationNumber) / InitialNumber;
        // ========================

        private readonly Graph graph;
        private readonly int number_of_cars;
        private readonly int number_of_shops;
        private List<Solution> population;
        private readonly int[] global_shops;
        private readonly int[] distance_from_storage_to;
        private readonly Random rand;

        public MtspSolver(Graph graph, int[] global_shops, int storage,
                int[] distance_from_storage_to, int number_of_cars)
        {
            rand = new Random();
            population = new List<Solution>();

            this.graph = graph;
            this.global_shops = global_shops;
            this.number_of_cars = number_of_cars;
            this.distance_from_storage_to = distance_from_storage_to;
            number_of_shops = graph.GetNumberOfShops();
        }

        public void Solve()
        {
            int cycles = 0;

            GenerateInitialPopulation();

            int i = 0;
            int best_result = 0;
            int last_best_result = 0;
            Solution best_solution;
            do {
                Mutation();
                Crossover();
                Selection(out best_solution, out best_result);

                if (last_best_result != best_result) {
                    cycles += i;
                    i = 0;
                    last_best_result = best_result;
                    Console.WriteLine(best_result);
                } else {
                    ++i;
                }
            } while (i < MaximumSameResults);
            cycles += i;
            
            Console.WriteLine(best_result + ": " + best_solution.GetGlobalString(global_shops));
            Console.WriteLine("Cycles: " + cycles);
        }

        private int ObjectiveFunction(Solution solution)
        {
            int maximum_distance = 0;
            foreach (int[] car_path in solution) {
                int distance = distance_from_storage_to[car_path[0]];
                int previous_shop = car_path[0];
                for (int i = 1; i < car_path.Length; ++i) {
                    distance += graph.GetWeight(previous_shop, car_path[i]);
                    previous_shop = car_path[i];
                }
                distance += distance_from_storage_to[previous_shop];
                if (distance > maximum_distance) {
                    maximum_distance = distance;
                }
            }
            return maximum_distance;
        }

        private void GenerateInitialPopulation()
        {
            for (int i = 0; i < InitialNumber; ++i) {
                this.population.Add(new Solution(number_of_shops, number_of_cars));
            }
        }

        private void Crossover()
        {
            if (number_of_shops < 2) {
                return;
            }

            for (int number = 0; number < CrossoverNumber; ++number) {
                int first_chromosome = rand.Next(0, population.Count);
                int second_chromosome;
                do {
                    second_chromosome = rand.Next(0, population.Count);
                } while (first_chromosome == second_chromosome);
                Solution first = population[first_chromosome];
                Solution second = population[second_chromosome];


                int[] second_shops_array = second.GetShopsArray();
                Solution child = first.Copy();
                int[] child_shops_array = child.GetShopsArray();
                bool[] is_position_assigned = new bool[number_of_shops];
                bool[] is_shop_assigned = new bool[number_of_shops];


                int length = rand.Next(1, number_of_shops - 1);
                int start_with = rand.Next(1, number_of_shops - length);
                for (int i = start_with; i < length + start_with; ++i) {
                    is_position_assigned[i] = true;
                    is_shop_assigned[child_shops_array[i]] = true;
                }

                for (int i = 0; i < number_of_shops; ++i) {
                    if (is_position_assigned[i]) {
                        continue;
                    }
                    
                    foreach (int second_shop in second_shops_array) {
                        if (!is_shop_assigned[second_shop]) {
                            child_shops_array[i] = second_shop;
                            is_position_assigned[i] = true;
                            is_shop_assigned[second_shop] = true;
                            break;
                        }
                    }
                }
                population.Add(child);
            }
        }

        private void Mutation()
        {
            for (int number = 0; number < MutationNumber; ++number) {
                int chromosome = rand.Next(0, population.Count);

                Solution new_solution = population[chromosome].Copy();
                
                int chance = rand.Next(0, 2);
                
                if (chance == 1) {
                    MutateSwap(new_solution);
                } else {
                    MutateAdjust(new_solution);
                }
                population.Add(new_solution);
            }
        }

        private void MutateSwap(Solution solution)
        {
            int[] shops = solution.GetShopsArray();
            int first_shop = rand.Next(0, number_of_shops);
            int second_shop = rand.Next(0, number_of_shops);
            while (first_shop == second_shop) {
                second_shop = rand.Next(0, number_of_shops);
            }
            int x = shops[first_shop];
            shops[first_shop] = shops[second_shop];
            shops[second_shop] = x;
        }

        private void MutateAdjust(Solution solution)
        {
            int[] cars = solution.GetPathLengths();
            int first_car = rand.Next(0, number_of_cars);
            int second_car = rand.Next(0, number_of_cars);
            while (first_car == second_car) {
                second_car = rand.Next(0, number_of_cars);
            }

            if (cars[first_car] == 1) {
                return;
            }
            int adjustment = rand.Next(1, cars[first_car] - 1);
            cars[first_car] -= adjustment;
            cars[second_car] += adjustment;
        }

        private void Selection(out Solution best_solution, out int best_function_result)
        {
            List<Solution>[] groups = new List<Solution>[InitialNumber];
            bool[] is_assigned = new bool[population.Count];
            for (int i = 0; i < population.Count; ++i) {
                int position = Global.GetRandomFree(is_assigned, population.Count - i);
                if (i % SelectionGroupSize == 0) {
                    groups[i / SelectionGroupSize] = new List<Solution>();
                }
                groups[i / SelectionGroupSize].Add(population[position]);
                is_assigned[position] = true;
            }

            List<Solution> new_population = new List<Solution>();
            List<int> function_results = new List<int>();
            
            foreach(List<Solution> group in groups)
            {
                int minimum_index = 0;
                int minimum_function_result = ObjectiveFunction(group[0]);
                for (int i = 1; i < group.Count; ++i)
                {
                    int function_result = ObjectiveFunction(group[i]);
                    if (minimum_function_result > function_result)
                    {
                        minimum_function_result = function_result;
                        minimum_index = i;
                    }
                }
                new_population.Add(group[minimum_index]);
                function_results.Add(minimum_function_result);
            }
            population = new_population;
            int minimum = function_results[0];
            int index = 0;
            for (int i = 1; i < population.Count; ++i) {
                if (function_results[i] < minimum) {
                    minimum = function_results[i];
                    index = i;
                }
            }
            best_solution = population[index].Copy();
            best_function_result = function_results[index];
        }
    }
}

File: src/mtsp_solver.cc
#include "mtsp_solver.h"
#include <iostream>

namespace mtsp {

void MtspSolver::importAdjacencyList(const AdjacencyList &adjacency_list,
                                     std::vector<std::size_t> &vertices,
                                     std::size_t origin,
                                     std::size_t number_of_chains) {
    _number_of_vertices = vertices.size();
    _number_of_chains = number_of_chains;

    _global_vertices = vertices;

    std::cout << "Converting the graph... " << std::endl;

    _graph.resize(_number_of_vertices);
    for (std::size_t u = 0; u < _number_of_vertices; ++u) {
        std::unordered_map<std::size_t, std::size_t> distance, previous;
        std::tie(distance, previous) = mtsp::dijkstra(adjacency_list,
                                                      _global_vertices.at(u));

        for (std::size_t v = u + 1; v < _number_of_vertices; ++v) {
            _graph.setWeight(u, v, distance.at(_global_vertices.at(v)));
        }

        _graph.setWeight(u, u, std::numeric_limits<unsigned long>::max());
    }
    std::cout << "Creating an origin distance vector" << std::endl;
    std::unordered_map<std::size_t, std::size_t> distance, previous;
    std::tie(distance, previous) = mtsp::dijkstra(adjacency_list, origin);
    _origin_distance.resize(_number_of_vertices);
    for (std::size_t u = 0; u < _number_of_vertices; ++u) {
        _origin_distance.at(u) = distance.at(_global_vertices.at(u));
    }

    std::cout << "Done" << std::endl;
}

void MtspSolver::solve() {
    unsigned cycles = 0;

    generateInitialPopulation();

    unsigned i = 0;
    unsigned long best_result = 0;
    unsigned long last_best_result = 0;
    std::size_t best_solution_index = 0;
    while (i < MaximumSameResults) {
        mutation();
        crossover();
        std::tie(best_solution_index, best_result) = selection();

        if (last_best_result != best_result) {
            cycles += i;
            i = 0;
            last_best_result = best_result;
            std::cout << best_result << '\n';
        } else {
            ++i;
        }
    }
    cycles += i;

    std::cout << "Best result: " << best_result << ". Vertices: ";
    for (std::size_t vertex: _population.at(best_solution_index)
                                                .getVertices()) {
        std::cout << vertex << " ";
    }
    std::cout << "\nCycles: " << cycles << std::endl;
}

const CompleteGraph &MtspSolver::getInternalGraph() const {
    return _graph;
}

unsigned long MtspSolver::objectiveFunction(const Solution &solution) {
    unsigned long maximum_distance = 0;
    std::size_t chain_start = 0;
    const std::vector<std::size_t> &vertices = solution.getVertices();
    for (std::size_t chain_length : solution.getChainsLengths()) {
        std::size_t previous_vertex = vertices.at(chain_start);
        unsigned long distance = _origin_distance.at(previous_vertex);
        for (std::size_t position = chain_start + 1;
                position != chain_start + chain_length;
                ++position) {
            distance += _graph.getWeight(previous_vertex,
                                         vertices.at(position));
            previous_vertex = vertices.at(position);
        }
        distance += _origin_distance.at(previous_vertex);
        if (distance > maximum_distance) {
            maximum_distance = distance;
        }
        chain_start += chain_length;
    }
    return maximum_distance;
}

void MtspSolver::generateInitialPopulation() {
    for (unsigned i = 0; i < InitialNumber; ++i) {
        _population.emplace_back(_number_of_vertices, _number_of_chains);
    }
}

void MtspSolver::crossover() {
    if (_number_of_vertices < 2 || _population.size() < 2) {
        return;
    }

    for (unsigned number = 0; number < CrossoverNumber; ++number) {
        auto first_chromosome = getRandomInteger<std::size_t>(0,
                                                _population.size());
        auto second_chromosome = getRandomInteger<std::size_t>(0,
                                                _population.size() - 1);
        if (second_chromosome >= first_chromosome) {
            ++second_chromosome;
        }
        Solution &first = _population.at(first_chromosome);
        Solution &second = _population.at(second_chromosome);

        _population.push_back(Solution::crossover(first, second));
    }
}

void MtspSolver::mutation() {
    int more_than_one_chain = 0;
    if (_number_of_chains >= 2) {
        more_than_one_chain = 1;
    }

    for (unsigned number = 0; number < MutationNumber; ++number) {
        auto chromosome = getRandomInteger<std::size_t>(0,
                                          _population.size());
        Solution new_solution(_population.at(chromosome).clone());

        int chance = more_than_one_chain * getRandomInteger(0, 2);

        if (chance == 0) {
            new_solution.mutateSwap();
        } else {
            new_solution.mutateAdjust();
        }
        _population.push_back(std::move(new_solution));
    }
}

std::tuple<std::size_t, unsigned long> MtspSolver::selection() {
    std::vector<std::vector<Solution>> groups(InitialNumber);
    std::vector<unsigned> is_assigned(_population.size(), 0);
    for (std::size_t i = 0; i < _population.size(); ++i) {
        std::size_t position = firstFreeRandom(is_assigned,
                                              _population.size() - i);
        groups.at(i / SelectionGroupSize).push_back(_population.at(position));
        is_assigned.at(position) = 1;
    }

    std::vector<Solution> new_population;
    std::vector<unsigned long> function_results;

    for (std::vector<Solution> &group : groups) {
        std::size_t minimum_index = 0;
        unsigned long minimum_function_result = objectiveFunction(group.at(0));
        for (unsigned i = 1; i < group.size(); ++i) {
            unsigned long function_result = objectiveFunction(group.at(i));
            if (minimum_function_result > function_result) {
                minimum_function_result = function_result;
                minimum_index = i;
            }
        }
        new_population.push_back(group.at(minimum_index));
        function_results.push_back(minimum_function_result);
    }

    _population.swap(new_population);

    unsigned long minimum = function_results.at(0);
    std::size_t index = 0;
    for (std::size_t i = 0; i < _population.size(); ++i) {
        if (function_results.at(i) < minimum) {
            minimum = function_results.at(i);
            index = i;
        }
    }

    return std::make_tuple(index, function_results.at(index));
}

}

